onnxruntime/core/external_ops/pyop.cc

```diff
@@ -0,0 +1,220 @@
+                                    std::multiplies<int64_t>());
+    auto np_array = reinterpret_cast<PyArrayObject*>(pyObj);
+    memcpy(PyArray_DATA(np_array), data, data_len);
+    return pyObj;
+}
+
+bool ExtractOutput(PyObject*                   pyObj,
+                   vector<unique_ptr<char[]>>& outputs,
+                   vector<int32_t>&            outputs_elem_size,
+                   vector<vector<int64_t>>&    outputs_dim)
+{
+    if (!PyArray_Check(pyObj)) {
+        return false;
+    }
+
+    outputs_dim.push_back({});
+    auto np_array = reinterpret_cast<PyArrayObject*>(pyObj);
+    outputs_elem_size.push_back(static_cast<int32_t>(PyArray_ITEMSIZE(np_array)));
+
+    for (int i = 0; i < PyArray_NDIM(np_array); ++i) {
+        outputs_dim.back().push_back(PyArray_SHAPE(np_array)[i]);
+    }
+
+    auto data_len = std::accumulate(begin(outputs_dim.back()),
+                                    end(outputs_dim.back()),
+                                    static_cast<int64_t>(outputs_elem_size.back()),
+                                    std::multiplies<int64_t>());
+
+    outputs.push_back(unique_ptr<char[]>(new char[data_len]));
+    memcpy(static_cast<void*>(outputs.back().get()), PyArray_DATA(np_array), data_len);
+    return true;
+}
+
+PYOP_EXPORT void* NewInstance(const char* module, const char* class_name, const unordered_map<string, string>& args)
+{
+    Scope scope; 
+    auto pyModule = PyImport_ImportModule(module);
+    if (nullptr == pyModule) {
+        return nullptr;
+    }
+
+    scope.Add(pyModule);
+    auto pyClass  = PyObject_GetAttrString(pyModule, class_name);
+    if (nullptr == pyClass) {
+        return nullptr;
+    }
+
+    scope.Add(pyClass);
+    auto empty_args = PyTuple_New(0);
+    scope.Add(empty_args);
+    auto named_args = PyDict_New();
+    scope.Add(named_args);
+    for (const auto& iter: args) {
+        PyDict_SetItemString(named_args, iter.first.c_str(), PyUnicode_FromString(iter.second.c_str()));
+    }
+
+    return PyObject_Call(pyClass, empty_args, named_args);
+}
+
+PYOP_EXPORT void ReleaseInstance(void* instance)
+{
+    Scope scope({static_cast<PyObject*>(instance)});
+}
+
+PYOP_EXPORT bool InvokePythonFunc(void*                            raw_inst,
+                                  const char*                      function,
+                                  const vector<const void*>&       inputs,
+                                  const vector<int32_t>&           inputs_type,
+                                  const vector<vector<int64_t>>&   inputs_dim,
+                                  vector<unique_ptr<char[]>>&      outputs,
+                                  vector<int32_t>&                 outputs_elem_size,
+                                  vector<vector<int64_t>>&         outputs_dim,
+                                  std::function<void(const char*)> logging_func = [](const char*){})
+{
+    Scope scope;
```

Comment:
This will cause only one python kernel to be executed at a time. Right? Let's document this as an explicit limitation.

Reply:
Yes let's do it.

Reply:
Will create a task for documentation.